### PR TITLE
Use cascade summation in nll_loss on CPU

### LINKED ISSUE #55657: `nn.CrossEntropyLoss` reductions for large CPUTensors yield wrong results
## 🐛 Bug

The loss value calculated by `nn.CrossEntropyLoss` via `reduction='mean'` vs. a manual mean calculation yields different results for "large" input tensors on the CPU.
Small inputs and CUDA seems to work fine.

The issue was reported in [the forum](https://discuss.pytorch.org/t/inconsistent-behaviour-of-cross-entropy-loss-without-weights/117540) by [Dominika_Basaj](https://discuss.pytorch.org/u/Dominika_Basaj) and [KFrank](https://discuss.pytorch.org/u/KFrank) created a great executable code snippet to isolate it down a bit more.
Thanks a lot to both of you!

## To Reproduce

Code snippet taken from [here](https://discuss.pytorch.org/t/inconsistent-behaviour-of-cross-entropy-loss-without-weights/117540/2?u=ptrblck).
```python
import torch
print (torch.__version__)

_ = torch.manual_seed (2021)

nBatch = 128
nClass = 2
height = 768
width = 768

input = torch.randn (nBatch, nClass, height, width)
target = torch.randint (nClass, (nBatch, height, width))

print ('nBatch =', nBatch, '\nnClass =', nClass, '\nheight =', height, '\nwidth =', width)
print ('big, float:')
print ('mean:', torch.nn.CrossEntropyLoss() (input, target))
print ('none:', torch.nn.CrossEntropyLoss (reduction = 'none') (input, target).mean())
print ('sum:', torch.nn.CrossEntropyLoss (reduction = 'sum') (input, target) / target.numel())

print ('big, cuda:')
print ('mean:', torch.nn.CrossEntropyLoss() (input.cuda(), target.cuda()))
print ('none:', torch.nn.CrossEntropyLoss (reduction = 'none') (input.cuda(), target.cuda()).mean())
print ('sum:', torch.nn.CrossEntropyLoss (reduction = 'sum') (input.cuda(), target.cuda()) / target.numel())

print ('big, double:')
print ('mean:', torch.nn.CrossEntropyLoss() (input.double(), target))
print ('none:', torch.nn.CrossEntropyLoss (reduction = 'none') (input.double(), target).mean())
print ('sum:', torch.nn.CrossEntropyLoss (reduction = 'sum') (input.double(), target) / target.numel())


nBatch = 8
nClass = 2
height = 16
width = 16

input = torch.randn (nBatch, nClass, height, width)
target = torch.randint (nClass, (nBatch, height, width))

print ('nBatch =', nBatch, '\nnClass =', nClass, '\nheight =', height, '\nwidth =', width)
print ('small, float:')
print ('mean:', torch.nn.CrossEntropyLoss() (input, target))
print ('none:', torch.nn.CrossEntropyLoss (reduction = 'none') (input, target).mean())
print ('sum:', torch.nn.CrossEntropyLoss (reduction = 'sum') (input, target) / target.numel())

print ('small, cuda:')
print ('mean:', torch.nn.CrossEntropyLoss() (input.cuda(), target.cuda()))
print ('none:', torch.nn.CrossEntropyLoss (reduction = 'none') (input.cuda(), target.cuda()).mean())
print ('sum:', torch.nn.CrossEntropyLoss (reduction = 'sum') (input.cuda(), target.cuda()) / target.numel())

print ('small, double:')
print ('mean:', torch.nn.CrossEntropyLoss() (input.double(), target))
print ('none:', torch.nn.CrossEntropyLoss (reduction = 'none') (input.double(), target).mean())
print ('sum:', torch.nn.CrossEntropyLoss (reduction = 'sum') (input.double(), target) / target.numel())
```

Output:
```python
1.8.1
nBatch = 128
nClass = 2
height = 768
width = 768
big, float:
mean: tensor(2.7632) # !!
none: tensor(0.9028) # !!
sum: tensor(0.6140) # !!
big, cuda:
mean: tensor(0.9028, device='cuda:0')
none: tensor(0.9028, device='cuda:0')
sum: tensor(0.9028, device='cuda:0')
big, double:
mean: tensor(0.9028, dtype=torch.float64)
none: tensor(0.9028, dtype=torch.float64)
sum: tensor(0.9028, dtype=torch.float64)
nBatch = 8
nClass = 2
height = 16
width = 16
small, float:
mean: tensor(0.9148)
none: tensor(0.9148)
sum: tensor(0.9148)
small, cuda:
mean: tensor(0.9148, device='cuda:0')
none: tensor(0.9148, device='cuda:0')
sum: tensor(0.9148, device='cuda:0')
small, double:
mean: tensor(0.9148, dtype=torch.float64)
none: tensor(0.9148, dtype=torch.float64)
sum: tensor(0.9148, dtype=torch.float64)
```

## Expected behavior

The calculated loss should be approx. equal up to the floating point precision.

## Environment

```python
PyTorch version: 1.8.1
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.2 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.19.6

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: A100-SXM4-40GB
GPU 1: A100-SXM4-40GB
GPU 2: A100-SXM4-40GB
GPU 3: A100-SXM4-40GB
GPU 4: A100-SXM4-40GB
GPU 5: A100-SXM4-40GB
GPU 6: A100-SXM4-40GB
GPU 7: A100-SXM4-40GB

Nvidia driver version: 450.51.06
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.2.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.2.0
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] torch==1.8.1
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               11.1.1               h6406543_8    conda-forge
[conda] mkl                       2020.4             h726a3e6_304    conda-forge
[conda] mkl-service               2.3.0            py38h1e0a361_2    conda-forge
[conda] mkl_fft                   1.3.0            py38h5c078b8_1    conda-forge
[conda] mkl_random                1.2.0            py38hc5bc63f_1    conda-forge
[conda] numpy                     1.19.2           py38h54aff64_0
[conda] numpy-base                1.19.2           py38hfa32c7d_0
[conda] pytorch                   1.8.1           py3.8_cuda11.1_cudnn8.0.5_0    pytorch
```

## Additional information

I've tried to set the number of interop and intraop threads to 1 as a quick debugging step, but this didn't help.

EDIT: Also same issue in the latest nightly: `1.9.0.dev20210408`

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @albanD @mruberry

### PR DESCRIPTION
**Pull request by peterbell10**
Fixes #55657

This also avoids summing `total_weight_val` when weights aren't supplied. Avoiding accumulated error completely.

### CODE REVIEW

File: aten/src/ATen/native/LossNLL.cpp [hunk 1]
```diff
@@ -82,43 +95,75 @@ static void nll_loss_out_frame(
...
+  constexpr int64_t cascade_sum_num_levels = 8;
+  scalar_t weight_partial_sums[cascade_sum_num_levels] = {0};
+  scalar_t loss_partial_sums[cascade_sum_num_levels] = {0};
+  const int64_t level_power =
+      std::max(int64_t(4), ceil_log2(numiter) / cascade_sum_num_levels);
+  const int64_t level_step = (1 << level_power);
+  const int64_t level_mask = level_step - 1;
+
+  for (int64_t b = 0; b < batch_size; b++) {
+    for (int64_t elem = 0; elem < map_size; elem++) {
```
Comment (ngimel) [inline aten/src/ATen/native/LossNLL.cpp:114]: why is there a loop over map_size (which is n_classes)? Will it lead to same data summed multiple times in lines 129-135?
Previously there was a loop over batch size only.

Comment (peterbell10) [reply aten/src/ATen/native/LossNLL.cpp]: Fixed and test are all passing again. Thanks!

File: aten/src/ATen/native/LossNLL.cpp [hunk 2]
```diff
@@ -23,6 +24,18 @@ inline scalar_t* optional_data(const Tensor& source) {
   return source.defined() ? source.data_ptr<scalar_t>() : nullptr;
 }
 
+
+static int64_t ceil_log2(int64_t x) {
```
Comment (ngimel) [inline aten/src/ATen/native/LossNLL.cpp:28]: given that it's used in a lot of places (and will be used in GroupNorm also, #54291), does it make sense to move it to some header and inline? #54291 suggests using native/cpu/utils.h.

**Review by ngimel — COMMENTED** [review]
This looks great, I have 1 small comment.

**Review by ngimel — APPROVED** [review]

File: aten/src/ATen/native/LossNLL.cpp [hunk 3]
```diff
@@ -82,41 +84,69 @@ static void nll_loss_out_frame(
...
+      weight_partial_sums[j + 1] += weight_partial_sums[j];
+      loss_partial_sums[j + 1] += loss_partial_sums[j];
+
+      weight_partial_sums[j] = 0;
+      loss_partial_sums[j] = 0;
+    }
+  }
+
+  if (weight_data) {
+    for (int64_t level = 0; level < cascade_sum_num_levels; ++level) {
```
Comment (ngimel) [inline aten/src/ATen/native/LossNLL.cpp:141]: nit: our linter wants
```
for (auto & weight_partial_sum : weight_partial_sums) {
```
, alternatively, you can use std::accumulate
here and in similar places.

Comment (peterbell10) [reply aten/src/ATen/native/LossNLL.cpp]: @ngimel changed to `std::accumulate`.
